File: tests/metrics/metric_pipeline_test.py
from __future__ import annotations
import numpy as np
import pytest
from taflow.metrics import MetricPipeline, SharpeRatio, SortinoRatio

def test_pipeline_computes_configured_metrics_under_caller_names() -> None:
    returns = np.array([0.01, -0.02, 0.03, 0.0, 0.04])
    pipeline = MetricPipeline()
    assert pipeline.add('sharpe', SharpeRatio()) is pipeline
    assert pipeline.add('sor', SortinoRatio()) is pipeline
    assert pipeline.from_returns(returns) is pipeline
    values = pipeline.compute()
    assert pipeline.metrics == ('sharpe', 'sor')
    assert values['sharpe'] == pytest.approx(SharpeRatio().from_returns(returns).compute())
    assert values['sor'] == pytest.approx(SortinoRatio().from_returns(returns).compute())

def test_pipeline_scalar_chunk_reset_and_domains_are_invariant() -> None:
    returns = np.array([0.1, -0.2, 0.05])
    batch = MetricPipeline().add('sharpe', SharpeRatio(12.0, 0.04))
    batch.from_returns(returns)
    scalar = MetricPipeline().add('sharpe', SharpeRatio(12.0, 0.04))
    scalar.from_returns([]).append(returns[0]).extend(returns[1:])
    assert scalar.compute() == pytest.approx(batch.compute())
    assert scalar.value == pytest.approx(batch.compute())
    assert scalar.reset().extend(returns).compute() == pytest.approx(batch.compute())
    log_returns = MetricPipeline().add('sharpe', SharpeRatio(12.0, 0.04))
    assert log_returns.from_log_returns(np.log1p(returns)).compute() == pytest.approx(batch.compute())
    equity = MetricPipeline().add('sharpe', SharpeRatio(12.0, 0.04))
    assert equity.from_equity([100.0, 110.0, 88.0, 92.4]).compute() == pytest.approx(batch.compute())
    pnl = MetricPipeline().add('sharpe', SharpeRatio(12.0, 0.04))
    assert pnl.from_pnl([10.0, -22.0, 4.4], 100.0).compute() == pytest.approx(batch.compute())

def test_pipeline_validation() -> None:
    pipeline = MetricPipeline().add('sharpe', SharpeRatio())
    with pytest.raises(ValueError):
        pipeline.add('sharpe', SortinoRatio())
    with pytest.raises(ValueError):
        pipeline.append(0.01)
    pipeline.from_returns([0.01])
    with pytest.raises(ValueError):
        pipeline.from_equity([100.0])
    with pytest.raises(ValueError):
        pipeline.add('late', SortinoRatio())

File: crates/taflow-metrics/src/metric_pipeline_test.rs
use super::{MetricPipeline, NanPolicy};
use crate::metrics::{SharpeRatio, SortinoRatio};

#[test]
fn configured_metrics_are_computed_under_caller_names() {
    let returns = [0.01, -0.02, 0.03, 0.0, 0.04];
    let mut pipeline = MetricPipeline::new();
    pipeline
        .add(
            "sharpe",
            SharpeRatio::new(252.0, 0.0, NanPolicy::Omit).unwrap(),
        )
        .unwrap()
        .add(
            "sor",
            SortinoRatio::new(252.0, 0.0, NanPolicy::Omit).unwrap(),
        )
        .unwrap()
        .from_returns(&returns)
        .unwrap();

    let values = pipeline.compute();
    assert_eq!(values[0].0, "sharpe");
    assert_eq!(values[1].0, "sor");
    assert!(values[0].1.is_some());
    assert!(values[1].1.is_some());
}

#[test]
fn scalar_chunk_reset_and_pnl_lifecycle_are_invariant() {
    let returns = [0.10, -0.20, 0.05];
    let mut batch = MetricPipeline::new();
    batch
        .add(
            "sharpe",
            SharpeRatio::new(12.0, 0.04, NanPolicy::Omit).unwrap(),
        )
        .unwrap()
        .from_returns(&returns)
        .unwrap();
    let expected = batch.compute()[0].1;

    let mut scalar = MetricPipeline::new();
    scalar
        .add(
            "sharpe",
            SharpeRatio::new(12.0, 0.04, NanPolicy::Omit).unwrap(),
        )
        .unwrap()
        .from_returns(&[])
        .unwrap()
        .append(returns[0])
        .unwrap()
        .extend(&returns[1..])
        .unwrap();
    assert_eq!(scalar.compute()[0].1, expected);
    scalar.reset().extend(&returns).unwrap();
    assert_eq!(scalar.compute()[0].1, expected);

    let mut pnl = MetricPipeline::new();
    pnl.add(
        "sharpe",
        SharpeRatio::new(12.0, 0.04, NanPolicy::Omit).unwrap(),
    )
    .unwrap()
    .from_pnl(&[10.0, -22.0, 4.4], 100.0)
    .unwrap();
    assert_eq!(pnl.compute()[0].1, expected);
}

#[test]
fn validates_names_order_and_domain_selection() {
    let mut pipeline = MetricPipeline::new();
    pipeline
        .add(
            "sharpe",
            SharpeRatio::new(252.0, 0.0, NanPolicy::Omit).unwrap(),
        )
        .unwrap();
    assert_eq!(pipeline.metric_names(), ["sharpe"]);
    assert!(pipeline
        .add(
            "sharpe",
            SortinoRatio::new(252.0, 0.0, NanPolicy::Omit).unwrap(),
        )
        .is_err());
    assert!(pipeline.append(0.01).is_err());
    pipeline.from_returns(&[0.01]).unwrap();
    assert!(pipeline.from_equity(&[100.0]).is_err());
    assert!(pipeline
        .add(
            "late",
            SortinoRatio::new(252.0, 0.0, NanPolicy::Omit).unwrap(),
        )
        .is_err());
}
